### ISSUE TITLE
error fetching tasks for application - Comparison method violates its general contract

### ISSUE THREAD

**Issue by emptywee** [spinnaker/spinnaker#5549]
### Issue Summary:

An exception is thrown when tasks are requested for an application where tasks with startTime=null exist. Orca responds with 500. 

### Cloud Provider(s):
Any?

### Environment:
Spinnaker 1.17.6

### Feature Area:
Orca

### Description:

500 Internal Server error on requesting `/gate/applications/app_name_here/tasks?statuses=` under certain circumstances (as outlined in the issue summary)

The following error is printed in orca's logs:
```
spin-orca-2 orca 2020-03-06 16:39:58.513 ERROR 1 --- [0.0-8083-exec-5] c.n.s.k.w.e.GenericExceptionHandlers     : [username] Internal Server Error
spin-orca-2 orca java.lang.IllegalArgumentException: Comparison method violates its general contract!
spin-orca-2 orca 	at java.util.TimSort.mergeHi(TimSort.java:899)
spin-orca-2 orca 	at java.util.TimSort.mergeAt(TimSort.java:516)
spin-orca-2 orca 	at java.util.TimSort.mergeForceCollapse(TimSort.java:457)
spin-orca-2 orca 	at java.util.TimSort.sort(TimSort.java:254)
spin-orca-2 orca 	at java.util.Arrays.sort(Arrays.java:1512)
spin-orca-2 orca 	at java.util.ArrayList.sort(ArrayList.java:1462)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.pipeline.persistence.jedis.RedisExecutionRepository.retrieveOrchestrationsForApplication(RedisExecutionRepository.java:642)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.telemetry.RedisInstrumentedExecutionRepository$retrieveOrchestrationsForApplication$2.invoke(RedisInstrumentedExecutionRepository.kt:179)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.telemetry.RedisInstrumentedExecutionRepository$retrieveOrchestrationsForApplication$2.invoke(RedisInstrumentedExecutionRepository.kt:31)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.telemetry.RedisInstrumentedExecutionRepository.withMetrics(RedisInstrumentedExecutionRepository.kt:45)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.telemetry.RedisInstrumentedExecutionRepository.retrieveOrchestrationsForApplication(RedisInstrumentedExecutionRepository.kt:178)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.pipeline.persistence.ExecutionRepository$retrieveOrchestrationsForApplication.call(Unknown Source)
spin-orca-2 orca 	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
spin-orca-2 orca 	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
spin-orca-2 orca 	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:143)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.controllers.TaskController.list(TaskController.groovy:118)
spin-orca-2 orca 	at com.netflix.spinnaker.orca.controllers.TaskController$$FastClassBySpringCGLIB$$7254ba71.invoke(<generated>)
```

### Steps to Reproduce:

The app has multiple tasks from 2018 and they were canceled, they have no startTime, they only have buildTime and endTime, e.g.:
```
{
  "id": "369055d2-294f-41f9-9345-0175b6faddef",
  "name": "Delete load balancer: employeeauth",
  "application": "employeeauth",
  "status": "CANCELED",
  "variables": [
    ...
  ],
  "steps": [
    ...
  ],
  "buildTime": 1516046458630,
  "endTime": 1516046766513,
  "execution": {
    "type": "ORCHESTRATION",
    "id": "369055d2-294f-41f9-9345-0175b6faddef",
    "application": "employeeauth",
    "buildTime": 1516046458630,
    "canceled": true,
    "canceledBy": "username",
    "limitConcurrent": false,
    "keepWaitingPipelines": false,
    "stages": [
      ...
    ],
    "endTime": 1516046766513,
    "status": "CANCELED",
    "authentication": {
      ...
    },
    "origin": "deck",
    "trigger": {
      ...
    },
    "description": "Delete load balancer: employeeauth",
    "notifications": [
      ...
    ],
    "initialConfig": {
      ...
    },
    "systemNotifications": [
      ...
    ]
  }
}
```

### Additional Details:

I patched orca's `orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java` at line 641 as follows:
```
    if (sorter != null) {
//      log.info("Sorting executions for application ({}) with sorter ({}) and criteria ({})", application, sorter, criteria);
      executions.forEach(exec -> {
        if (exec.getStartTime() == null) {
          exec.setStartTime(0L);
        }
      });
      try {
        executions.sort(sorter);
      } catch (IllegalArgumentException exc) {
        log.error("Could not sort executions for application ({})!", application);
        exc.printStackTrace();
      }
    }
```
And I was able to get response and get most of the tasks sorted correctly. Those without the startTime field went to the end of the list and probably remained unsorted. This unblocked me, but I think the sorter/comparator should be fixed.

**Comment by ezimanyi** [comment]
This was fixed in spinnaker/orca#3392.  I've cherry-picked to 1.17 and 1.18 in spinnaker/orca#3498 and spinnaker/orca#3499.

**Comment by emptywee** [comment]
Thanks a bunch, @ezimanyi !
